### ISSUE TITLE
Restoring a partition of a deleted service

### ISSUE THREAD

**Issue by mm-set** [Microsoft/service-fabric#206]
I'm trying to use the BackupRestoreService to restore a partition from a previously deleted service.
Let's say I have to recreate my whole cluster and I have existing backups from an old cluster. Now I want to restore the backups to the new cluster. Of course, the IDs of the partitions are now different.
I have tried the following: 

- Get the latest backups of the deleted service
- Match min/max keys of the partitions of the old service to the partitions of the new service
- Call partition restore API with the backup information I got from the previous step (BackupId, BackupLocation and BackupStorage details)

With this, I just get the response 
`{"Error":{"Code":"E_UNEXPECTED","Message":"Unable to get recovery point details."}}`
from the restore partition API call

Is this scenario even possible with the BackupRestoreService? Is there any other approach to restore partitions if a cluster needs to be recreated?

**Comment by VipulM-MSFT** [comment]
It should be possible - adding @raunakpandya for follow up.

**Comment by raunakpandya** [comment]
Yes. This is definitely possible and infact one of the prime use case for restore. Can you share more details on how are you invoking the restore API, parameters etc? Adding @hrushib

**Comment by mm-set** [comment]
First, I'm using the GetBackups API to retrieve the latest backups of an application/service:
```
$url = "https://oldcluster.northeurope.cloudapp.azure.com:19080/Applications/myApp/$/GetBackups?api-version=6.2-preview&Latest=true"

$response = Invoke-WebRequest -Uri $url -Method Get -CertificateThumbprint '***'

$BackupPoints = (ConvertFrom-Json $response.Content)
$BackupPoints.Items
```
From these results, I match the old partition ID and new partition ID with the min or max key from the partition info. Partitioning scheme did not change, of course. 

```
BackupId                : a0eac05d-202d-429e-b610-93bf11a86d44
BackupChainId           : 8bee7aeb-053f-4e3e-8668-4a8c0d761f4e
ApplicationName         : fabric:/myApp
ServiceName             : fabric:/myApp/myService
PartitionInformation    : @{LowKey=3689348814741910325; HighKey=5534023222112865485; ServicePartitionKind=Int64Range; Id=cef63ff6-1fd9-4bd2-8b67-b5f6209c9e3d}
BackupLocation          : myApp\myService\cef63ff6-1fd9-4bd2-8b67-b5f6209c9e3d\2018-08-28 07.50.33.zip
BackupType              : Incremental
EpochOfLastBackupRecord : @{DataLossNumber=131795785510236564; ConfigurationNumber=42949672960}
LsnOfLastBackupRecord   : 100777
CreationTimeUtc         : 2018-08-28T07:50:33Z
FailureError            :
```

With this information, I call the Restore API:

```
$BackupStorageDescription = @{
    StorageKind = 'AzureBlobStore'
    ConnectionString = '***'
    ContainerName = 'backup-container'
}

$RestorePartitionDescription = @{
    BackupId = 'a0eac05d-202d-429e-b610-93bf11a86d44'
    BackupLocation = 'myApp\\myService\\cef63ff6-1fd9-4bd2-8b67-b5f6209c9e3d\\2018-08-28 07.50.33.zip'
    BackupStorage = $BackupStorageDescription
}

$body = (ConvertTo-Json $RestorePartitionDescription)
$url = "https://newcluster.northeurope.cloudapp.azure.com:19080/Partitions/1cf8a539-c3c0-43df-9101-d8bccab32de5/$/Restore?api-version=6.2-preview"

Invoke-WebRequest -Uri $url -Method Post -Body $body -ContentType 'application/json' -CertificateThumbprint '***'
```

Whereas `1cf8a539-c3c0-43df-9101-d8bccab32de5` is the partition ID of the new partition I want to restore.

From the restore API call, I get this response:  
`{"Error":{"Code":"E_UNEXPECTED","Message":"Unable to get recovery point details."}}`

Both, old and new cluster have the same backup policy and configuration, pointing to the same storage account and blob container.

**Comment by raunakpandya** [comment]
Thanks for the details. The flow is all correct. Seems like some issue in trying to fetch the backup chain which needs to be restored here. Can you share your cluster log details (cluster resource ID and region, assuming its northeurope resion) over email (raunakp at Microsoft dot com)?

**Comment by mm-set** [comment]
Unfortunately, I already deleted the cluster. I'm currently in the process of setting up a new one. I will try to reproduce my described problem and send you the details via email once I'm done.

**Comment by raunakpandya** [comment]
Yes, please do share. We can quickly take a look and unblock you here.

**Comment by mm-set** [comment]
Setup of the new cluster is now completed and had the same issue again.
I've sent you an email with detail information.

**Comment by hrushib** [comment]
@mm-set, please use single path separator instead of double in `BackupLocation` value as depicted below.

```PowerShell
$RestorePartitionDescription = @{
    BackupId = 'a0eac05d-202d-429e-b610-93bf11a86d44'
    BackupLocation = 'myApp\myService\cef63ff6-1fd9-4bd2-8b67-b5f6209c9e3d\2018-08-28 07.50.33.zip'
    BackupStorage = $BackupStorageDescription
}
```

Let me know if this results in successful restore.

**Comment by mm-set** [comment]
Thanks. That was the problem.
Now the API call was successful. Unfortunately, no data was restored. I saw in SF Explorer that the partition was shortly in quorum loss, then back to healthy. 
The backups I have for this partition is 1 full backup (~50MB), followed by 3 incremental backups (each ~8MB). I restored from the last incremental backup. How long should the restore procedure take for 1 partition?

**Comment by aagup** [comment]
@mm-set The restore is getting blocked as the files (backups) which are downloaded by Backup Restore  Service for restore are being locked by some anti-virus software (mostly defender). Hence the restore request is failing. 

The real reason can be seen from EventViewer logs or the logs generated by Service Fabric.

I will request you to kindly retry the request. 
OR 
Mark the Data folder for Service fabric in exception list for the anti - virus.

**Comment by mm-set** [comment]
@aagup I couldn't find any logs in event viewer regarding blocked files by anti-virus. I disabled windows defender anyways on the node of the primary replica, but it didn't help.

**Comment by hrushib** [comment]
@mm-set, we have identified root cause of the issue. There is bug in the backup copier path which leads to premature download operation completion. Fix for this issue will be released soon. Thanks for your help and patience.

**Comment by raunakpandya** [comment]
@mm-set - This issue is resolved as part of Service Fabric v6.4 release. I am going to close this now.
